pkg/tlsutil/tls.go

```diff
@@ -0,0 +1,113 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tlsutil
+
+import (
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// CertType defines the type of the cert.
+type CertType int
+
+const (
+	// ClientCert defines a client cert.
+	ClientCert CertType = iota
+	// ServingCert defines a serving cert.
+	ServingCert
+	// ClientAndServingCert defines both client and serving cert.
+	ClientAndServingCert
+)
+
+// CertConfig configures how to generate the Cert.
+type CertConfig struct {
+	// CertName is the name of the cert.
+	CertName string
+	// Optional CertType. Serving, client or both; defaults to both.
+	CertType CertType
+	// Optional CommonName is the common name of the cert; defaults to "".
+	CommonName string
+	// Optional Organization is Organization of the cert; defaults to "".
+	Organization []string
+	// Optional CA Key, if user wants to provide custom CA key via a file path.
+	CAKey string
+	// Optional CA Certificate, if user wants to provide custom CA cert via file path.
+	CACert string
+	// TODO: consider to add passed in SAN fields.
+}
+
+// CertGenerator is an operator specific TLS tool that generates TLS assets for the deploying a user's application.
+type CertGenerator interface {
+	// GenerateCert generates a secret containing TLS encryption key and cert, a Secret
+	// containing the CA key, and a ConfigMap containing the CA Certificate given the Custom
+	// Resource(CR) "cr", the Kubernetes Service "Service", and the CertConfig "config".
+	//
+	// GenerateCert creates and manages TLS key and cert and CA with the following:
+	// CA creation and management:
+	// - If CA is not given:
+	//  - A unique CA is generated for the CR.
+	//  - CA's key is packaged into a Secret as shown below.
+	//  - CA's cert is packaged in a ConfigMap as shown below.
+	//  - The CA Secret and ConfigMap are created on the k8s cluster in the CR's namespace before
+	//    returned to the user. The CertGenerator manages the CA Secret and ConfigMap to ensure it's
+	//    unqiue per CR.
+	// - If CA is given:
+	//  - CA's key is packaged into a Secret as shown below.
+	//  - CA's cert is packaged in a ConfigMap as shown below.
+	//  - The CA Secret and ConfigMap are returned but not created in the K8s cluster in the CR's
+	//    namespace. The CertGenerator doesn't manage the CA because the user controls the lifecycle
+	//    of the CA.
+	//
+	// TLS Key and Cert Creation and Management:
+	// - A unique TLS cert and key pair is generated per CR + CertConfig.CertName.
+	// - We use the CA determined from the above to generate and to sign the cert of the TLS key.
```

Comment:
Since the CA is always used(user specified or newly generated) how about we make this more concise:
```
The CA is used to generate and sign the TLS cert.
```
Either that or reword `CA determined from the above` since it doesn't quite sound right to me.

Reply:
yeah, not a problem. I was just trying to be more specific on where the CA is from.